plexus-compilers/plexus-compiler-aspectj/src/main/java/org/codehaus/plexus/compiler/ajc/AspectJCompiler.java

```diff
@@ -557,83 +564,157 @@ private void checkForAspectJRT( List<String> cp )
-        }
-        else if ( "1.7".equals( sourceVersion ) )
-        {
-            buildConfig.getOptions().sourceLevel = ClassFileConstants.JDK1_7;
-        }
-        else if ( "1.6".equals( sourceVersion ) )
-        {
-            buildConfig.getOptions().sourceLevel = ClassFileConstants.JDK1_6;
-        }
-        else if ( "1.5".equals( sourceVersion ) )
+        buildConfig.getOptions().sourceLevel = versionStringToMajorMinor( sourceVersion );
+    }
+
+    /**
+     * Set the target version in AspectJ compiler
+     *
+     * @param buildConfig
+     * @param targetVersion
+     */
+    private void setTargetVersion( AjBuildConfig buildConfig, String targetVersion )
+        throws CompilerException
+    {
+        buildConfig.getOptions().targetJDK = versionStringToMajorMinor( targetVersion );
+    }
+
+    private static long versionStringToMajorMinor(String version) throws CompilerException
+    {
+        if ( version == null )
+        {
+            version = "";
+        }
+        // Note: We avoid using org.codehaus.plexus:plexus-java here on purpose, because Maven Compiler might depend on
+        // a different (older) versionm, e.g. not having the 'asMajor' method yet. This can cause problems for users
+        // trying to compile their AspectJ code using Plexus.
+        
+        version = version.trim()
+            // Cut off leading "1.", focusing on the Java major
+            .replaceFirst( "^1[.]", "" )
+            // Accept, but cut off trailing ".0", as ECJ/ACJ explicitly support versions like 5.0, 8.0, 11.0
+            .replaceFirst("[.]0$", "");
+
+        switch ( version )
+        {
+            // Java 1.6 as a default source/target seems to make sense. Maven Compiler should set its own default
+            // anyway, so this probably never needs to be used. But not having a default feels bad, too.
+            case "" : return ClassFileConstants.JDK1_6;
+            case "1" : return ClassFileConstants.JDK1_1;
+            case "2" : return ClassFileConstants.JDK1_2;
+            case "3" : return ClassFileConstants.JDK1_3;
+            case "4" : return ClassFileConstants.JDK1_4;
+            case "5" : return ClassFileConstants.JDK1_5;
+            case "6" : return ClassFileConstants.JDK1_6;
+            case "7" : return ClassFileConstants.JDK1_7;
+            case "8" : return ClassFileConstants.JDK1_8;
+            case "9" : return ClassFileConstants.JDK9;
+            case "10" : return ClassFileConstants.JDK10;
+            case "11" : return ClassFileConstants.JDK11;
+            case "12" : return ClassFileConstants.JDK12;
+            case "13" : return ClassFileConstants.JDK13;
+            case "14" : return ClassFileConstants.JDK14;
+            case "15" : return ClassFileConstants.JDK15;
+            case "16" : return ClassFileConstants.JDK16;
+        }
+        throw new CompilerException( "Unknown Java source/target version number: " + version );
```

Comment:
I changed the error message in order to reflect that unlike before, here it is not always about a source version but could also be a target version (or compliance level, more generally).